File: offer/simple/PHP/exchange.php
<?php

/**
 * 剑指 Offer 21. 调整数组顺序使奇数位于偶数前面
 * @Author: pengwei
 * @Date:   2021-05-20 09:48:24
 * @Last Modified by:   pengwei
 */
function exchange(array $nums)
{
	// 方法1-暴力法
	// $oddNum  = [];
	// $evenNum = [];

	// // 通过位与运算符判断奇偶性
	// foreach ($nums as $v) {
	// 	if ($v & 1) {
	// 		$oddNum[] = $v;
	// 	} else {
	// 		$evenNum[] = $v;
	// 	}
	// }

	// // 合并
	// foreach ($evenNum as $v) {
	// 	$oddNum[] = $v;
	// }
	// return $oddNum;
	

	// 方法2-双指针法
	$left  = 0;
	$right = count($nums) - 1;

	while ($left < $right) {
		// 左边移动-为奇数
		if ($nums[$left] & 1) {
			++$left;
			continue;
		}

		// 右边移动-为偶数
		if (($nums[$right] & 1) == 0) {
			--$right;
			continue;
		}

		$tmp = $nums[$left];
		$nums[$left] = $nums[$right];
		$nums[$right] = $tmp;;
	}
	return $nums;
}

File: offer/simple/Golang/exchange.go
package main

import "fmt"

// 输入一个整数数组，实现一个函数来调整该数组中数字的顺序，使得所有奇数在数组的前半部分，所有偶数在数组的后半部分。
//
//示例：
//输入：nums = [1,2,3,4]
//输出：[1,3,2,4]
//注：[3,1,2,4] 也是正确的答案之一。

func main() {

	nums := []int {1, 2, 3, 4, 5, 6}
	fmt.Println(exchange(nums))
}

func exchange(nums []int) []int {

	len := len(nums)
	if len < 2 {
		return nums
	}

	// 1 2 3 4 5
	i := 0
	j := len - 1
	for i < j {
		if nums[i] & 1 == 1 {
			i++
			continue
		}

		if nums[j] & 1 == 0 {
			j--
			continue
		}

		tmp := nums[i]
		nums[i] = nums[j]
		nums[j] = tmp
	}
	return nums
}
